src/REL/d/a/obj/d_a_obj_block_underground.cpp

```diff
@@ -1,6 +1,292 @@
+#include "d/snd/d_snd_small_effect_mgr.h"
+#include "f/f_base.h"
+#include "f/f_list_nd.h"
+#include "m/m_vec.h"
+#include "nw4r/g3d/res/g3d_resfile.h"
+#include "nw4r/g3d/res/g3d_resmdl.h"
+#include "toBeSorted/d_emitter.h"
+
 SPECIAL_ACTOR_PROFILE(OBJ_BLOCK_UNDERGROUND, dAcOBlockUnderground, fProfile::OBJ_BLOCK_UNDERGROUND, 0x13F, 0, 3);
 
-dAcOBlockUnderground::dAcOBlockUnderground() {}
-dAcOBlockUnderground::~dAcOBlockUnderground() {}
+const s32 dAcOBlockUnderground::sSubtypeIdx[12] = {0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 1, 1};
+
+const char *const dAcOBlockUnderground::sOarcName = "MoleBlock";
+
+const char *const dAcOBlockUnderground::sMdlNames[4] = {
+    "MoleBlock",
+    "MoleBlockBomb",
+    "MoleBlockBreak",
+    "MoleBlockBreak",
+};
+
+const char *const dAcOBlockUnderground::sDzbPlcNames[4] = {
+    "MoleBlock_rock",
+    "MoleBlock_rock",
+    "MoleBlock_tuti",
+    "MoleBlock_tuti",
+};
+
+const dCcD_SrcCyl dAcOBlockUnderground::sCylSrc = {
+    /* mObjInf */
+    {/* mObjAt */ {0, 0, {0, 0, 0}, 0, 0, 0, 0, 0, 0},
+     /* mObjTg */ {AT_TYPE_BOMB, 0x11, {0, 8, 0x407}, 0x0, 0x0},
+     /* mObjCo */ {0}},
+    /* mCylInf */
+    {50.f, 100.f}
+};
+
+u32 dAcOBlockUnderground::getCylTgType() {
+    return 0x20020;
+}
+
+const char *dAcOBlockUnderground::getOarcName() {
+    return sOarcName;
+}
+
+extern "C" cListMg_c *fn_8018B840(); // d_underground
+
+dAcOBlockUnderground::dAcOBlockUnderground() : mRef(this), mUndergroundRef(nullptr) {
+    fn_8018B840()->append(&mRef);
+}
+
+dAcOBlockUnderground::~dAcOBlockUnderground() {
+    fn_8018B840()->remove(&mRef);
+    delete mCyl;
+}
+
+bool dAcOBlockUnderground::createHeap() {
+    const char *oarcName = getOarcName();
+    void *data = getOarcResFile(oarcName);
+    if (data == nullptr) {
+        return false;
+    }
+
+    nw4r::g3d::ResFile resFile(data);
+    if (!resFile.IsValid()) {
+        return false;
+    }
+
+    const char *mdlName = getMdlName();
+    nw4r::g3d::ResMdl mdl = resFile.GetResMdl(mdlName);
+    if (!mdl.IsValid()) {
+        return false;
+    }
+
+    if (!mMdl.create(mdl, &heap_allocator, 0x120, 1, nullptr)) {
+        return false;
+    }
+
+    void *dzb = getOarcDZB(getOarcName(), sDzbPlcNames[getSubtypeIdx()]);
+    if (dzb == nullptr) {
+        return false;
+    }
+
+    void *plc = getOarcPLC(getOarcName(), sDzbPlcNames[getSubtypeIdx()]);
+    if (plc == nullptr) {
+        return false;
+    }
+
+    bool result = !mBgW.Set((cBgD_t *)dzb, (PLC *)plc, cBgW::MOVE_BG_e, &mWorldMtx, &mScale);
+    if (!result) {
+        return false;
+    }
+
+    if (fn_459_A00()) {
+        dCcD_Cyl *cyl = new dCcD_Cyl();
+        mCyl = cyl;
+        if (mCyl == nullptr) {
+            return false;
+        }
+    }
+    return true;
+}
+
+int dAcOBlockUnderground::actorCreate() {
+    u8 sceneFlag = getSecondSceneFlag();
+    if (sceneFlag < 0xff) {
+        bool flag = SceneflagManager::sInstance->checkFlag(roomid, sceneFlag);
+        if (flag) {
```

Comment:
```suggestion
        if (SceneflagManager::sInstance->checkBoolFlag(roomid, sceneFlag)) {
```
should help eliminating the temp